components/gitpod-db/src/typeorm/auth-code-repository-db.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { log } from '@gitpod/gitpod-protocol/lib/util/logging';
+import { DateInterval, OAuthAuthCode, OAuthAuthCodeRepository, OAuthClient, OAuthScope, OAuthUser } from "@jmondi/oauth2-server";
+import * as crypto from 'crypto';
+import { inject, injectable } from "inversify";
+import { EntityManager, Repository } from "typeorm";
+import { DBOAuthAuthCodeEntry } from './entity/db-oauth-auth-code';
+import { TypeORM } from './typeorm';
+
+const expiryInFuture = new DateInterval("5m");
+
+@injectable()
+export class AuthCodeRepositoryDB implements OAuthAuthCodeRepository {
+
+    @inject(TypeORM)
+    private readonly typeORM: TypeORM;
+
+    protected async getEntityManager(): Promise<EntityManager> {
+        return (await this.typeORM.getConnection()).manager;
+    }
+
+    async getOauthAuthCodeRepo(): Promise<Repository<DBOAuthAuthCodeEntry>> {
+        return (await this.getEntityManager()).getRepository<DBOAuthAuthCodeEntry>(DBOAuthAuthCodeEntry);
+    }
+
+    public async getByIdentifier(authCodeCode: string): Promise<OAuthAuthCode> {
+        log.info(`getByIdentifier ${authCodeCode}`);
+        const authCodeRepo = await this.getOauthAuthCodeRepo();
+        let authCodes = await authCodeRepo.find({ code: authCodeCode });
+        log.info(`getByIdentifier pre: ${JSON.stringify(authCodes)}`);
+        authCodes = authCodes.filter(te => (new Date(te.expiresAt)).getTime() > Date.now());
+        log.info(`getByIdentifier post: ${JSON.stringify(authCodes)}`);
+        const authCode = authCodes.length > 0 ? authCodes[0] : undefined;
+        if (!authCode) {
+            throw new Error(`authentication code not found`);
+        }
+        return authCode;
+    }
+    public issueAuthCode(client: OAuthClient, user: OAuthUser | undefined, scopes: OAuthScope[]): OAuthAuthCode {
+        const code = crypto.randomBytes(30).toString('hex');
+        log.info(`issueAuthCode: ${JSON.stringify(client)}, ${JSON.stringify(user)}, ${JSON.stringify(scopes)}, ${code}`);
+        // NOTE: caller (@jmondi/oauth2-server) is responsible for adding the remaining items, PKCE params, redirect URL, etc
```

Comment:
That requires moving the entire AuthCodeRepositoryDB (OAuthAuthCodeRepository) implementation to server and since it requires storage, then adding a db service + implementation for it to use.  

I agree it meets the current, very structured implementation approach for *most* parts of the system that interact with the db and will take a look at it tomorrow.

Reply:
> That requires moving the entire AuthCodeRepositoryDB (OAuthAuthCodeRepository) implementation to server

no it doesn't. 

this pattern is used in may cases here. the `DB` implementation lives in `gitpod-db` and is injected into services from `server`.

Reply:
> odd to implement the missing bits for the framework within gitpod-db.

Not so much a missing bit but the OAuthAuthCodeRepository interface needs to be implemented for the framework to use. Moving 1 requires moving the rest (yes - there are other, less readable ways to do so).

> this pattern is used in may cases here. the `DB` implementation lives in `gitpod-db` and is injected into services from `server`.

Agreed. That's what I was referring to with "then adding a db service + implementation for it to use"